operator/sidecar/tbot.go

```diff
@@ -0,0 +1,254 @@
+package sidecar
+
+import (
+	"context"
+	"fmt"
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/auth/authclient"
+	"github.com/gravitational/teleport/lib/tbot"
+	"github.com/gravitational/teleport/lib/tbot/config"
+	"github.com/gravitational/teleport/lib/tbot/identity"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	"time"
+)
+
+const (
+	DefaultCertificateTTL  = 2 * time.Hour
+	DefaultRenewalInterval = 30 * time.Minute
+)
+
+// ClientAccessor returns a working teleport auth client when invoked.
+// Client users should always call this function on a regular basis to ensure certs are always valid.
+type ClientAccessor func(ctx context.Context) (auth.ClientI, error)
+
+// Bot is a wrapper around an embedded tbot.
+// It implements sigs.k8s.io/controller-runtime/manager.Runnable and
+// sigs.k8s.io/controller-runtime/manager.LeaderElectionRunnable so it can be added to a controllerruntime.Manager.
+type Bot struct {
+	cfg        *config.BotConfig
+	running    bool
+	rootClient auth.ClientI
+	opts       Options
+}
+
+func (b *Bot) initializeConfig() {
+	// Initialize the memory stores. They contain identities renewed by the bot
+	// We're reading certs directly from them
+	rootMemoryStore := &config.DestinationMemory{}
+	destMemoryStore := &config.DestinationMemory{}
+
+	// Initialize tbot config
+	b.cfg = &config.BotConfig{
+		Onboarding: &config.OnboardingConfig{
+			Token:      "",         // Field should be populated later, before running
+			CAPins:     []string{}, // Field should be populated later, before running
+			JoinMethod: types.JoinMethodToken,
+		},
+		Storage: &config.StorageConfig{
+			DestinationMixin: config.DestinationMixin{
+				Memory: rootMemoryStore,
+			},
+		},
+		Destinations: []*config.DestinationConfig{
+			{
+				DestinationMixin: config.DestinationMixin{
+					Memory: destMemoryStore,
+				},
+			},
+		},
+		Debug:           false,
+		AuthServer:      b.opts.Addr,
+		CertificateTTL:  DefaultCertificateTTL,
+		RenewalInterval: DefaultRenewalInterval,
+		Oneshot:         false,
+	}
+
+	_ = b.cfg.CheckAndSetDefaults()
+
+	for _, artifact := range identity.GetArtifacts() {
+		_ = destMemoryStore.Write(artifact.Key, []byte{})
+		_ = rootMemoryStore.Write(artifact.Key, []byte{})
+	}
+
+}
+
+func (b *Bot) GetClient(ctx context.Context) (auth.ClientI, error) {
+	if !b.running {
```

Comment:
Are you sure there's no risk of concurrent access to `b.running` ? Shouldn't this have a mutex on it ?